### PR TITLE
Bug 1926154: Fix for creating new pool with arbiter -  wrong replica

### PR DESCRIPTION
**Pull request by GowthamShanmugam**
Signed-off-by: Gowtham Shanmugasundaram <gshanmug@redhat.com>

### CODE REVIEW

File: frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx [hunk 1]
```diff
@@ -121,7 +122,25 @@ export const StoragePoolModal = withHandlePromise((props: StoragePoolModalProps)
...
+
+  React.useEffect(() => {
+    if (isArbiterCluster) {
+      setReplicaSize('4');
+    }
+  }, [isArbiterCluster]);
+
+  const replicaList: string[] = _.keys(OCS_DEVICE_REPLICA).filter(function(replica: string) {
+    return (isArbiterCluster && replica === '4') || (!isArbiterCluster && replica !== '4');
+  });
```
Comment (GowthamShanmugam) [inline frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx:141]: It is not mandatory to keep this (isArbiterCluster && replica === '4'), anyway we are hiding the list for the arbiter, But for safety, i am keeping this.

**Comment by GowthamShanmugam** [comment]
Arbiter SC

Arbiter_replica

**Comment by GowthamShanmugam** [comment]
Non-Arbiter SC
Non_arbiter

File: frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx [hunk 2]
```diff
@@ -121,7 +122,25 @@ export const StoragePoolModal = withHandlePromise((props: StoragePoolModalProps)
...
   }, [isSubmit, newPool, newPoolLoadError, newPoolLoaded, timer]);
 
-  const replicaDropdownItems = _.keys(OCS_DEVICE_REPLICA).map((replica) => {
+  const [storageCluster] = useK8sGet<ListKind<StorageClusterKind>>(
+    OCSServiceModel,
+    null,
+    CEPH_STORAGE_NAMESPACE,
+  );
+
+  const isArbiterCluster: boolean = storageCluster?.items[0]?.spec?.arbiter?.enable;
```
[the remark this answers is not in the document]
Comment (GowthamShanmugam) [reply frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx:131]: ack

**Comment by GowthamShanmugam** [comment]
@cloudbehl  changes done

File: frontend/packages/ceph-storage-plugin/src/components/modals/add-capacity-modal/add-capacity-modal.tsx [hunk 3]
```diff
@@ -54,8 +55,8 @@ export const AddCapacityModal = (props: AddCapacityModalProps) => {
   const isNoProvionerSC: boolean = storageClass?.provisioner === NO_PROVISIONER;
   const selectedSCName: string = getName(storageClass);
   const deviceSetIndex: number = getCurrentDeviceSetIndex(deviceSets, selectedSCName);
-  const hasFlexibleScaling = ocsConfig?.spec?.flexibleScaling;
-  const isArbiterEnabled: boolean = ocsConfig?.spec?.arbiter?.enable;
+  const hasFlexibleScaling = checkHasFlexibleScaling(ocsConfig);
+  const isArbiterEnabled: boolean = checkClusterIsArbiter(ocsConfig);
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/components/modals/add-capacity-modal/add-capacity-modal.tsx:59]:
```suggestion
  const hasFlexibleScaling = checkFlexibleScaling(ocsConfig);
  const isArbiterEnabled: boolean = checkArbiterCluster(ocsConfig);
```

File: frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx [hunk 4]
```diff
@@ -121,7 +124,22 @@ export const StoragePoolModal = withHandlePromise((props: StoragePoolModalProps)
     }
   }, [isSubmit, newPool, newPoolLoadError, newPoolLoaded, timer]);
 
-  const replicaDropdownItems = _.keys(OCS_DEVICE_REPLICA).map((replica) => {
+  const [storageCluster, storageClusterLoaded, storageClusterLoadError] = useK8sGet<
+    ListKind<StorageClusterKind>
+  >(OCSServiceModel, null, CEPH_STORAGE_NAMESPACE);
+
+  React.useEffect(() => {
+    setIsArbiterCluster(checkClusterIsArbiter(storageCluster?.items[0]));
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx:132]:
```suggestion
    setArbiterCluster(checkArbiterCluster(storageCluster?.items[0]));
```

File: frontend/packages/ceph-storage-plugin/src/utils/modal-common.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { StorageClusterKind } from '../types';
```
Comment (cloudbehl) [inline frontend/packages/ceph-storage-plugin/src/utils/modal-common.ts:1]: We can change the filename to `feature-common.ts` maybe.

File: frontend/packages/ceph-storage-plugin/src/components/modals/add-capacity-modal/add-capacity-modal.tsx (hunk 3 above)
Comment (GowthamShanmugam) [reply frontend/packages/ceph-storage-plugin/src/components/modals/add-capacity-modal/add-capacity-modal.tsx:59]: ack

File: frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx (hunk 4 above)
Comment (GowthamShanmugam) [reply frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx:132]: ack

File: frontend/packages/ceph-storage-plugin/src/utils/modal-common.ts (hunk 5 above)
Comment (GowthamShanmugam) [reply frontend/packages/ceph-storage-plugin/src/utils/modal-common.ts:1]: ack

File: frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx [hunk 6]
```diff
@@ -121,7 +124,22 @@ export const StoragePoolModal = withHandlePromise((props: StoragePoolModalProps)
...
+  const [storageCluster, storageClusterLoaded, storageClusterLoadError] = useK8sGet<
+    ListKind<StorageClusterKind>
+  >(OCSServiceModel, null, CEPH_STORAGE_NAMESPACE);
+
+  React.useEffect(() => {
+    setArbiterCluster(checkArbiterCluster(storageCluster?.items[0]));
+    if (isArbiterCluster) {
+      setReplicaSize('4');
+    }
+  }, [storageCluster, storageClusterLoaded, storageClusterLoadError, isArbiterCluster]);
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx:136]:
```suggestion
const storageClusterMemoized = useDeepCompareMemoize(storageCluster, true);
const isArbiterCluster = checkArbiterCluster(storageClusterMemoized);
```

File: frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx [hunk 7]
```diff
@@ -121,7 +124,22 @@ export const StoragePoolModal = withHandlePromise((props: StoragePoolModalProps)
...
+
+  const replicaList: string[] = _.keys(OCS_DEVICE_REPLICA).filter(function(replica: string) {
+    return (isArbiterCluster && replica === '4') || (!isArbiterCluster && replica !== '4');
+  });
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx:140]:
```suggestion
  const replicaList: string[] = _.keys(OCS_DEVICE_REPLICA).filter((replica: string) => (isArbiterCluster && replica === '4') || (!isArbiterCluster && replica !== '4')));
```

File: frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx [hunk 8]
```diff
@@ -121,7 +124,22 @@ export const StoragePoolModal = withHandlePromise((props: StoragePoolModalProps)
...
+
+  const replicaDropdownItems = replicaList.map((replica) => {
     return (
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx:143]:
```suggestion
  const replicaDropdownItems = replicaList.map((replica) => (
```

File: frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx [hunk 9]
```diff
@@ -84,6 +86,7 @@ export const StoragePoolModal = withHandlePromise((props: StoragePoolModalProps)
   /* TODO: use reducer */
   const [isSubmit, setIsSubmit] = React.useState(false);
   const [timer, setTimer] = React.useState<NodeJS.Timer>(null);
+  const [isArbiterCluster, setArbiterCluster] = React.useState(false);
```
Comment (bipuladh) [inline frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx:89]:
```suggestion
```

**Review by bipuladh — CHANGES_REQUESTED** [review]

File: frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx (hunk 6 above)
Comment (GowthamShanmugam) [reply frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx:136]: ack

Comment (GowthamShanmugam) [reply frontend/packages/ceph-storage-plugin/src/components/modals/storage-pool-modal/storage-pool-modal.tsx:136]: As we discussed, i am keeping this change as it is

**Comment by GowthamShanmugam** [comment]
@cloudbehl @bipuladh changes done

**Pull request merged**